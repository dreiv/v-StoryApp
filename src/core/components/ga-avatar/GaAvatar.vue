<script setup lang="ts">
export interface GaAvatarProps {
  size?: 'small' | 'medium' | 'large'
  content?: string
}

const { size = 'medium', content } = defineProps<GaAvatarProps>()
</script>

<template>
  <div :class="[$style.avatar, $style[size]]">
    {{ content }}
  </div>
</template>

<style module>
.avatar {
  display: inline-flex;
  justify-content: center;
  align-items: center;
  border-radius: var(--ga-radius-round);
  background-color: var(--ga-color-surface-action);
  color: var(--ga-color-text-on-action);
  font-weight: var(--ga-font-weight-medium);
  font-family: var(--ga-font-family-primary);
}

.small {
  width: var(--ga-size-spacing-06);
  height: var(--ga-size-spacing-06);
  font-size: var(--ga-text-sm-font-size);
}

.medium {
  width: var(--ga-size-spacing-08);
  height: var(--ga-size-spacing-08);
  font-size: var(--ga-text-md-font-size);
}

.large {
  width: var(--ga-size-spacing-10);
  height: var(--ga-size-spacing-10);
  font-size: var(--ga-text-lg-font-size);
}
</style>
