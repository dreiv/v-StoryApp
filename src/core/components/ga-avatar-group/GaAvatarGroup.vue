<script setup lang="ts">
import { computed, useCssModule } from 'vue'
import GaAvatar from '../ga-avatar/GaAvatar.vue'

export interface GaAvatarGroupProps {
  /**
   * Size of all avatars in the group
   * @default 'medium'
   */
  size?: 'extra-small' | 'small' | 'medium' | 'large'
  /**
   * Maximum number of avatars to display
   * @default undefined
   */
  max?: number
  /**
   * Amount of overlap between avatars (in pixels)
   * @default undefined - uses default overlap based on size
   */
  overlap?: number
  /**
   * Text for the excess avatar count when max is reached
   */
  overflowLabel?: string
}

const props = withDefaults(defineProps<GaAvatarGroupProps>(), {
  size: 'medium',
})

const style = useCssModule()

// Get avatar width based on size for proper half-overlap calculation
const getAvatarWidth = computed(() => {
  const avatarWidths = {
    'extra-small': 24, // estimated width in pixels
    small: 32,
    medium: 40,
    large: 48,
  }
  return avatarWidths[props.size]
})

// Calculate overlap values (when not explicitly provided)
const getDefaultOverlap = computed(() => {
  // Default to half the avatar width for natural looking overlap
  if (props.overlap !== undefined) return props.overlap

  // Calculate half of the avatar width for overlap
  return Math.round(getAvatarWidth.value / 2)
})

// Get border thickness based on size
const getBorderThickness = computed(() => {
  const borderValues = {
    'extra-small': 1,
    small: 1.5,
    medium: 2,
    large: 3,
  }
  return borderValues[props.size]
})
</script>

<template>
  <div
    :class="style.group"
    :style="{
      '--avatar-width': `${getAvatarWidth}px`,
      '--avatar-overlap': `${getDefaultOverlap}px`,
      '--border-thickness': `${getBorderThickness}px`,
    }"
  >
    <!-- Regular slot for child avatars -->
    <slot></slot>

    <!-- Overflow indicator -->
    <GaAvatar v-if="overflowLabel" :size="size" :content="overflowLabel" :class="style.count" />
  </div>
</template>

<style module>
.group {
  display: flex;
  flex-direction: row-reverse; /* Reverse order so last avatar appears on top */
  align-items: center;
  width: max-content;
}

.group :deep(.avatar) {
  position: relative;
  transition:
    transform 0.2s ease,
    box-shadow 0.2s ease;
  margin-left: calc(var(--avatar-overlap) * -1); /* Negative margin for overlap */
  border: var(--border-thickness) solid var(--ga-color-surface-primary);
}

/* When hovering/focusing, bring avatar forward with transform and box-shadow */
.group :deep(.avatar:hover),
.group :deep(.avatar:focus-visible) {
  transform: translateY(-4px) scale(1.05);
  z-index: 1; /* Only need one level of z-index here */
  box-shadow: 0 4px 8px rgba(0, 0, 0, 0.15);
}

/* Remove negative margin from first element in the reversed flexbox (actually the last element visually) */
.group :deep(.avatar:first-child) {
  margin-left: 0;
}

.count {
  background-color: var(--ga-color-surface-secondary);
}
</style>
