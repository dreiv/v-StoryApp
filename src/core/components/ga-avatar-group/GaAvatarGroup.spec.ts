import { mount, shallowMount, type VueWrapper } from '@vue/test-utils'
import GaAvatarGroup from './GaAvatarGroup.vue'
import GaAvatar from '../ga-avatar/GaAvatar.vue'

describe('GaAvatarGroup', () => {
  let wrapper: VueWrapper

  it('should render correctly with default props', () => {
    wrapper = shallowMount(GaAvatarGroup)
    expect(wrapper.html()).toMatchSnapshot()
  })

  describe('Props', () => {
    it('should apply correct CSS variables based on size', () => {
      wrapper = mount(GaAvatarGroup, { props: { size: 'medium' } })
      const style = wrapper.attributes('style')
      expect(style).toContain('--border-thickness: 2px')
    })

    it('should display overflow label when provided', () => {
      wrapper = mount(GaAvatarGroup, {
        props: {
          overflowLabel: '+3',
          size: 'medium',
        },
      })

      // Check if GaAvatar is rendered with the overflow label
      const avatars = wrapper.findAllComponents(GaAvatar)
      expect(avatars.length).toBe(1)
      expect(avatars[0].props('content')).toBe('+3')
    })
  })

  describe('Slot handling', () => {
    it('should render child avatars', () => {
      wrapper = mount(GaAvatarGroup, {
        slots: {
          default: `
            <GaAvatar content="AV" />
            <GaAvatar content="BN" />
          `,
        },
        global: {
          components: {
            GaAvatar,
          },
        },
      })

      const avatars = wrapper.findAllComponents(GaAvatar)
      expect(avatars.length).toBe(2)
    })

    it('should apply size prop to overflow label avatar', () => {
      wrapper = mount(GaAvatarGroup, {
        props: {
          overflowLabel: '+3',
          size: 'small',
        },
      })

      const avatar = wrapper.findComponent(GaAvatar)
      expect(avatar.props('size')).toBe('small')
    })
  })
})
