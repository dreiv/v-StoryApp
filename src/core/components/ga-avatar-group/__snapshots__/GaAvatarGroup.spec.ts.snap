// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GaAvatarGroup > should render correctly with default props 1`] = `
"<div class="group" style="--avatar-width: 40px; --avatar-overlap: 20px; --border-thickness: 2px;">
  <!-- Regular slot for child avatars -->
  <!-- Overflow indicator -->
  <!--v-if-->
</div>"
`;
